mozjs/build.rs

```diff
@@ -131,24 +138,102 @@ fn cc_flags() -> Vec<&'static str> {
 }
 
+fn cargo_target_dir() -> PathBuf {
+    let out_dir = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let mut dir = out_dir.as_path();
+    while let Some(target_dir) = dir.parent() {
+        if target_dir.file_name().unwrap().to_string_lossy() == "target" {
+            return target_dir.to_path_buf();
+        }
+        dir = target_dir;
+    }
+    panic!("$OUT_DIR is not in target")
+}
+
+fn find_moztools() -> Option<PathBuf> {
+    let cargo_target_dir = cargo_target_dir();
+    let deps_dir = cargo_target_dir.join("depedencies");
+    let moztools_path = deps_dir.join(MOZ_TOOLS);
+
+    if moztools_path.exists() {
+        Some(moztools_path)
+    } else {
+        None
+    }
+
+    // to support many moztools versions
+    /*
+    // maybe moztools are already in targed/depedencies/moztools-*
+    if let Ok(dirs) = std::fs::read_dir(deps_dir) {
+        dirs.into_iter().find_map(|p| {
+            if let Ok(pp) = p {
+                if let Some(name) = pp.path().file_name() {
+                    if name.to_string_lossy().contains("moztools") {
+                        Some(pp.path())
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        })
+    } else {
+        None
+    }*/
+}
+
 fn build_jsapi(build_dir: &Path) {
+    for (key, value) in env::vars() {
+        println!("{key}: {value}");
+    }
     let target = env::var("TARGET").unwrap();
     let mut make = find_make();
 
-    // If MOZILLA_BUILD is specified, process that environment variable to put
-    // the build tools on the path and properly set the AUTOCONF environment
-    // variable.
-    if let Some(mozilla_build_path) = env::var_os("MOZILLA_BUILD") {
+    #[cfg(windows)]
+    {
+        let moztools = if let Some(moztools) = env::var_os("MOZ_TOOLS_PATH") {
+            PathBuf::from(moztools)
+        } else if let Some(moztools) = find_moztools() {
+            // moztools already in targed/depedencies/moztools-*
+            moztools
+        } else {
+            // download moztools
+            let cargo_target_dir = cargo_target_dir();
+            let deps_dir = cargo_target_dir.join("depedencies");
+
+            // curl is on Win10+
+            let mut curl = Command::new("curl");
+
+            let moztools_zip = &deps_dir.join("moztools.zip");
+            let result = curl
+                .arg("-SL")
+                .arg(MOZ_TOOLS_DL)
+                .arg("--create-dirs")
+                .args(&["-o", &moztools_zip.display().to_string()])
+                .status()
+                .unwrap();
+            assert!(result.success());
+
+            // extract
+            let file = fs::File::open(moztools_zip).unwrap();
+            let mut archive = zip::ZipArchive::new(file).unwrap();
+            archive.extract(deps_dir).unwrap();
+
+            find_moztools().unwrap()
```

Comment:
The `.unwrap()`'s in these lines should be replaced with `.expect()` with an appropriate error message so errors can be diagnosed more easily.